plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -3613,6 +3614,55 @@ public void testEmptyDelete()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testReadFromVersionedTableWithSchemaEvolution()
+    {
+        String tableName = "test_versioned_table_schema_evolution_" + randomTableSuffix();
+
+        assertQuerySucceeds("CREATE TABLE " + tableName + "(col1 varchar)");
+        long v1SnapshotId = getLatestSnapshotId(tableName);
+        assertThat(query("SELECT * FROM " + tableName + " FOR VERSION AS OF " + v1SnapshotId))
+                .hasOutputTypes(ImmutableList.of(VARCHAR))
+                .returnsEmptyResult();
+
+        assertUpdate("ALTER TABLE " + tableName + " ADD COLUMN  col2 integer");
+        assertThat(query("SELECT * FROM " + tableName))
+                .hasOutputTypes(ImmutableList.of(VARCHAR, INTEGER))
+                .returnsEmptyResult();
+
+        assertUpdate("INSERT INTO " + tableName + " VALUES ('a', 11)", 1);
+        long v2SnapshotId = getLatestSnapshotId(tableName);
+        assertThat(query("SELECT * FROM " + tableName + " FOR VERSION AS OF " + v2SnapshotId))
+                .hasOutputTypes(ImmutableList.of(VARCHAR, INTEGER))
+                .matches("VALUES (CAST('a' AS varchar), 11)");
+        assertThat(query("SELECT * FROM " + tableName))
+                .hasOutputTypes(ImmutableList.of(VARCHAR, INTEGER))
+                .matches("VALUES (CAST('a' AS varchar), 11)");
+
+        assertUpdate("ALTER TABLE " + tableName + " ADD COLUMN  col3 bigint");
+        assertThat(query("SELECT * FROM " + tableName + " FOR VERSION AS OF " + v2SnapshotId))
```

Comment:
This is a bit peculiar - the same query 

```
"SELECT * FROM " + tableName + " FOR VERSION AS OF " + v2SnapshotId
```

returns now 3 columns, and a few lines later, after doing the `INSERT` it will return only 2 columns.

The reasoning for this peculiarity is that there are no snapshots done for structural changes (same as in Spark) so in case we're dealing with the latest snapshot of a table, we switch by default to `table.schema()` and not

```
int schemaId = table.snapshot(snapshotIdYouWant).schemaId(); 
Schema schema = table.schemas().get(schemaId);
```

which may point to an outdated schema for the table.

Reply:
Sorry if this is silly question but is it a correct behaviour that iceberg's creator are aware of ? Is it possible that it is made by mistake ? Sorry if this was discussed already

Reply:
I raised this comment as a discussion on the Trino project, but you are right - this topic actually affects the Apache Iceberg as well.

I started a Slack discussion on it:

https://apache-iceberg.slack.com/archives/C025PH0G1D4/p1654858193311349

Reply:
In case of a `CREATE TABLE` statement in Trino (I don't know if accidentally or intentionally) there is a snapshot being created. This is not the case for 

In order to avoid dealing with situations as the one exposed above, I'd advocate for committing an empty transaction.

```
            transaction.newFastAppend().commit();
            transaction.commitTransaction();
```

Reply:
> Another approach would be to not always fill `IcebergTableHandle.snapshotId` with `table.currentSnapshot().snapshotId()` in `io.trino.plugin.iceberg.IcebergMetadata#getSnapshotId` , but the logic of the `trino-iceberg` is currently highly dependendent on this field to be always filled.

We want to know which snapshots we're querying. I don't want the code to be dependent on some assumption that no new snapshot shows up, or be vulnerable to a race condition if it does